File: pathplode/trunk/pathplode.pl
#! /usr/bin/perl 
#
# Perl implementation of pathplode
#
# Copyright 2003 by Ulf Klaperski
#
# This file is part of pathplode.
# 
###############################################################################
#
# Pathplode is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
# 
# Pathplode is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with the pathplode sources; see the file COPYING.  
#
###############################################################################
#
# $Id$
#
###############################################################################

use Getopt::Long qw(:config no_ignore_case);

GetOptions ('help|h!' => \$help,
	    'uniquify|u!' => \$uniquify,
	    'unique|U!' => \$unique,
	    'separator|S' => \$separator,
	    'list|t!' => \$list,
	    'prepend|p=s' => \$prepend,
	    'append|a=s' => \$append,
	    'remove-first|f=s' => \$remove_first,
	    'remove-last|l=s' => \$remove_last,
	    'remove-all|d=s' => \$remove_all,
	    'after|A=s' => \$after,
	    'before|B=s' => \$before,
	    'allow-empty|E' => \$allow_empty,
	    'color|C' => \$color,
	    'preserve-trailing-slash|P' => \$preserve_trailing_slash
	   );


sub help {
  print "Usage: pathplode [options] command <element> <extra-arg> [path]
    Commands:
	--append  <element>: put element at end of list
	--prepend <element>: put element at beginning of list
	--remove-first <regexp>: remove the first matching element 
	--remove-all <regexp>: 
	--uniquify
	--list
    Options:
	--unique: remove multiple entries. Only the entry with the
    		  highest priority is left over.
	--after <regexp>: modify append/prepend to insert new element after
			  element matching regexp. Note that prepend inserts
			  after the first matching element, append after the last
			  matching element.
	--before <regexp>: modify append/prepend to insert new element before
			   element matching regexp. Note that prepend inserts
			   before the first matching element, append before the last
			   matching element.
	--separator: The separator between path elements. Default: ':'
	--preserve-trailing-slash: by default a trailing slash is removed. This changes
				   the behaviour to keep those slashes.
	--allow-empty: Allow empty entries. By default they are removed.
	--color: colorize output - currently only affects the list command.
    	
Examples:
	pathplode --uniquify ~/gnu/bin:/usr/local/bin:/usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games:/usr/local/bin
		  -> ~/gnu/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
	pathplode --unique --prepend ~/gnu/bin /usr/local/bin:/usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games
		  -> ~/gnu/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
	pathplode --append ~/gnu/bin /usr/local/bin:/usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games
		  -> /usr/local/bin:/usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games:~/gnu/bin
	pathplode --prepend ~/bin --after gnu /usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games:~/gnu/bin
		  -> /usr/bin:/bin:~/gnu/bin:~/bin:/usr/bin/X11:/usr/games:~/gnu/bin
";
  exit shift;
}

###############################################################################
# Commands
###############################################################################
# 
# Each command expects as first argument a reference to an array of the current  
# elements. The result will be passed back by modifying this array.

sub uniquify {
  my $elements_r = shift;
  my %found = ();
  local @new_elements = ();
  foreach $elm (@$elements_r) {
    next if $found{$elm};
    $found{$elm} = 1;
    push @new_elements, $elm;
  }
  @$elements_r = @new_elements;
}

sub prepend {
  my $prepend = shift;
  my $elements_r = shift;
  my ($found, $idx) = (0, 0);
  
  uniquify $elements_r if $unique_before;

  if ($before or $after) {
    my $anchor = ($before ? $before : $after);
    while (!$found and $idx<$#$elements_r) {
      if ($$elements_r[$idx] =~ /$anchor/) {
	$found=1;
	last;
      }
      $idx++;
    }
    
    unless ($found) {
      $idx=0;
      $error=1;
      print STDERR "Anchor element $anchor not found. Prepending at beginning!\n";
    } elsif ($after) {
      $idx++; 
    }
  }
  splice @$elements_r, $idx, 0, $prepend;
  
  uniquify $elements_r if $unique_after;
}

sub append {
  my $append = shift;
  my $elements_r = shift;
  my ($found, $idx) = (0, scalar @$elements_r);
  uniquify $elements_r if $unique_before;

  if ($before or $after) {
    $anchor = ($before ? $before : $after);
    while (!$found and $idx>0) {
      if ($$elements_r[$idx-1] =~ /$anchor/) {
	$found=1;
	last;
      }
      $idx--;
    }
    
    unless ($found) {
      $idx=@$elements;
      $error=1;
      print STDERR "Anchor element $anchor not found. Prepending at beginning!\n";
    } elsif ($before) {
      $idx--;
    }
  }
  splice @$elements_r, $idx, 0, $append;

  uniquify $elements_r if $unique_after;
}

sub remove_first {
  my $remove = shift;
  my $elements_r = shift;
  my ($found, $idx) = (0, 0);
  
  uniquify $elements_r if $unique_before;

  if ($before or $after) {
    $anchor = ($before ? $before : $after);
  } else {
    undef $anchor;
  }

  while (1) {
    if ($$elements_r[$idx] =~ /$remove/) {
      if ((!defined($before) or ($idx!=$#$elements_r and $$elements_r[$idx+1] =~ /$before/) ) 
	  and (!defined($after) or ($idx!=0 and $$elements_r[$idx-1] =~ /$after/))) {
	$found=1;
	last;
      }
    }
    last unless ( ++$idx<@$elements_r );
  }
  
  if ($found) {
    print "Removing $idx\n";
    splice @$elements_r, $idx, 1;
  }
  
  uniquify $elements_r if $unique_after;
}

sub remove_last {
  my $remove = shift;
  my $elements_r = shift;

  uniquify $elements_r if $unique_before;

  ($before, $after) = ($after, $before);
  @$elements_r = reverse @$elements_r;
  remove_first $remove, $elements_r;
  @$elements_r = reverse @$elements_r;

  uniquify $elements_r if $unique_after;

}

sub list {
  my $elements_r = shift;
  my %paths;
  foreach $path (@$elements_r) {
    print $color && $paths{$path}?"\033[1;31m":"","$path\n", $color?"\e[0m":"";
    $paths{$path}=1;
  }
}

###############################################################################
# Main Program
###############################################################################

$help and &help;

$path_in = pop;

defined $separator or $separator = ':';
@commands = ("uniquify",  "prepend", "append", "remove_first", "remove_last", "list");

foreach $cmd (@commands) {
  if ($$cmd) {
    if ($cmd_found) {
      die "Only one command of ", join(" ", @commands), " allowed!\n";
    }
    $cmd_found=1;
    $command=$cmd;
  }
}

if ($unique) {
  $unique_before = 1;
  $unique_after = 1;
}

@elements = split /$separator/, $path_in;
map { s:/$::; } @elements unless $preserve_trailing_slash;

if ($command eq "uniquify") {
  uniquify \@elements;
} elsif ($command eq "prepend") {
  $prepend =~ s:/$:: unless $preserve_trailing_slash;
  prepend $prepend, \@elements;
} elsif ($command eq "append") {
  $append =~ s:/$:: unless $preserve_trailing_slash;
  append $append, \@elements;
} elsif ($command eq "remove_first") {
  remove_first $remove_first, \@elements;
} elsif ($command eq "remove_last") {
  remove_last $remove_last, \@elements;
} elsif ($command eq "list") {
  list \@elements;
  $noprint=1;
}

print join $separator, @elements unless $noprint;

File: tags/release-0.6/pathplode.cc
/*! \file pathplode.cc
 * @author    Ulf Klaperski
 *
 * C++ implementation of pathplode
 *
 * Copyright 2004 by Ulf Klaperski
 *
 * This file is part of pathplode.
 * 
 *##############################################################################
 *
 * Pathplode is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 * 
 * Pathplode is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with the pathplode sources; see the file COPYING.  
 *
 *##############################################################################
 *
 * $Id$
 */

#include<iostream>
using namespace std;

extern char *optarg;
#include <getopt.h>
#include <queue>
#include "pathp_list.h"


#define VERSION "0.6"

#define USAGE "\n\
Usage: pathplode [options] <command> [options] <command> ... PATH\n\
Modify the incoming path as specified by the commands and options and print it to standard output.\n\
Information:\n\
  --help (-h)\n\
  --version (-v)\n\
Commands:\n\
  --append(-a) <element>: put element at end of list\n\
  --prepend(-p) <element>: put element at beginning of list\n\
  --remove-first(-f) <element>: remove the first matching element \n\
  --remove-last(-l) <element>: remove the last matching element \n\
  --remove-all(-d) <element>: remove all occurrences of element\n\
  --uniquify(-u): remove all multiple entries, just leaving the first one \n\
  --list(-t): List all path entries, each on its own line\n\
Local options:  \n\
  --reset-options(-r): reset options to their default values for the following commands.\n\
  --after(-A) <anchor>: modify append/prepend to insert new element after\n\
        the given element. Note that prepend inserts\n\
        after the first matching element, append after the last\n\
        matching element.\n\
  --before(-B) <anchor>: modify append/prepend to insert new element before\n\
         the given element. Note that prepend inserts\n\
         before the first matching element, append before the last\n\
         matching element.\n\
  --allow-empty(-E): Allow empty entries. By default they are removed.\n\
  --colour(-C): colorize output - currently only affects the list command.\n\
  --regex(-R): Use regular expressions. All hooks and elements to remove are treated\n\
         as regular expressions (GNU regex syntax).\n\
Global options:  \n\
  --preserve-trailing-slash(-P): by default a trailing slash is removed. This changes\n\
           the behaviour to keep those slashes.\n\
  --separator(-S): The separator between path elements. Default: ':'\n\
  --unique(-U): calls a uniquify at the beginning and at the end, just a shortcut for\n\
           putting -u twice. This ensures matching at the right place and leaving a\n\
           uniquified path.\n\
NOTE:\n\
  The order of options matters! Every option affects only subsequent commands, until\n\
  it is changed again. To remove all option use the --reset-options option.\n\
Examples:\n\
  pathplode --uniquify ~/gnu/bin:/usr/local/bin:/usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games:/usr/local/bin\n\
      -> ~/gnu/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games\n\
  pathplode --unique --prepend ~/gnu/bin /usr/local/bin:/usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games\n\
      -> ~/gnu/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games\n\
  pathplode --append ~/gnu/bin /usr/local/bin:/usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games\n\
      -> /usr/local/bin:/usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games:~/gnu/bin\n\
  pathplode --after ~/gnu/bin --prepend ~/bin /usr/bin:/bin:~/gnu/bin:/usr/bin/X11:/usr/games:~/gnu/bin\n\
      -> /usr/bin:/bin:~/gnu/bin:~/bin:/usr/bin/X11:/usr/games:~/gnu/bin\n\
  pathplode --regex --after 'X1[12]' --prepend ~/new /usr/bin:/bin:/usr/bin/X12:~/gnu/bin:/usr/bin/X11:/usr/games\n\
      -> /usr/bin:/bin:/usr/bin/X12:~/new:~/gnu/bin:/usr/bin/X11:/usr/games\n";


#define COPYING "\n\
Copyright (C) 2004, 2005 Ulf Klaperski.\n\
Pathplode comes with ABSOLUTELY NO WARRANTY.\n\
You may redistribute copies of Pathplode\n\
under the terms of the GNU General Public License.\n\
For more information about these matters, see the file named COPYING.";

static bool unique_opt = false;
static bool list_paths_occured = false;

enum pp_cmd {
  undefined,
  uniquify,
  remove_first,
  remove_last,
  remove_all,
  append,
  prepend,
  list_paths,
  before_elm,
  after_elm,
  colour,
  reset_options,
  use_regex
};

string version_string = string("Pathplode, version ") + VERSION;
const char usage_string[] = USAGE;
const char copying_string[] = COPYING;

// just a helper class to store command and argument in a list
class cmd_with_arg {
public:
  cmd_with_arg(pp_cmd command, string argument) {
    cmd=command;
    arg=argument;
  }
  pp_cmd cmd;
  string arg;
};

static queue<cmd_with_arg> commands;

void process_options (int* argc, char** argv[], pp_cmd &command, string &command_arg,
                      pathp_list &pplist) {
  int c;
  int option_index = 0;

  static struct option pathplode_options[] =
  {
    {"help", 0, 0, 'h'},
    {"version", 0, 0, 'v'},
    {"list", 0, 0, 't'},
    {"uniquify", 0, 0, 'u'},
    {"prepend", 1, 0, 'p'},
    {"append", 1, 0, 'a'},
    {"reset-options", 0, 0, 'r'},
    {"remove-first", 1, 0, 'f'},
    {"remove-last", 1, 0, 'l'},
    {"remove-all", 1, 0, 'd'},
    {"after", 1, 0, 'A'},
    {"before", 1, 0, 'B'},
    {"colour", 0, 0, 'C'},
    {"allow-empty", 0, 0, 'E'},
    {"separator", 1, 0, 'S'},
    {"unique", 0, 0, 'U'},
    {"preserve-trailing-slash", 0, 0, 'P'},
    {"regex", 0, 0, 'R'},
    {0, 0, 0, 0}
  };

  while (1)
  {
    c = getopt_long (*argc, *argv, "hvtup:a:f:l:d:A:B:CrRES:UP",
                     pathplode_options, &option_index);
    if (c<0) break;

    switch (c)
    {
      case 'h':
        cout << version_string << endl;
        cout << usage_string << endl;
        cout << copying_string << endl;
        exit(0);
      case 'v':
        cout << version_string << endl;
        cout << copying_string << endl;
        exit(0);
      case 'u':
        commands.push(cmd_with_arg(uniquify, ""));
        break;
      case 'p':
        commands.push(cmd_with_arg(prepend, optarg));
        command = prepend;
        command_arg = optarg;
        break;
      case 'a':
        commands.push(cmd_with_arg(append, optarg));
        break;
      case 'f':
        commands.push(cmd_with_arg(remove_first, optarg));
        break;
      case 'l':
        commands.push(cmd_with_arg(remove_last, optarg));
        break;
      case 'd':
        commands.push(cmd_with_arg(remove_all, optarg));
        break;
      case 't':
        commands.push(cmd_with_arg(list_paths, optarg));
        break;
      case 'A':
        commands.push(cmd_with_arg(after_elm, optarg));
        break;
      case 'E':
        pplist.set_allow_empty(true);
        break;
      case 'U':
        unique_opt = true;
        break;
      case 'B':
        commands.push(cmd_with_arg(before_elm, optarg));
        break;
      case 'R':
        commands.push(cmd_with_arg(use_regex, ""));
        break;
      case 'r':
        commands.push(cmd_with_arg(reset_options, ""));
        break;
      case 'C':
        commands.push(cmd_with_arg(colour, optarg));
        break;
      case 'P':
        pplist.set_purge_trailing_slash(false);
        break;
      case 'S':
        pplist.set_separator(optarg[0]);
        break;
        
      default:
        cout << c << " HHHHMMM..." << endl;
    }
  }
}

/*!  \fn int main(int argc, char* argv[], char *env[]) 
 * \brief Anzeige der Aufrufparameter des Programms und der Umgebungsvariablen.

 * Die folgenden Parameter werden vom Betriebsystem übergeben:
  \param argc  Anzahl der Argumente
  \param argv[] C-String-Array der weiteren Argumente
  \param env[] C-String-Array der Umgebungsvariablen
*/
int main(int argc, char* argv[], char *env[]) {
  pp_cmd command = undefined;
  string command_arg = "";
  
  string path_list_in; 
  pathp_list all_paths("");
  int error_occured = 0;
  process_options(&argc, &argv, command, command_arg, all_paths);
  if (argc>0) path_list_in = argv[argc-1];
  all_paths.set_list_from_string(path_list_in);

  if (unique_opt) error_occured |= all_paths.uniquify();

  while(!commands.empty()) {
    cmd_with_arg &next_command = commands.front();
    switch (next_command.cmd) {
    case uniquify:
      error_occured |= all_paths.uniquify();
      break;
    case list_paths:
      error_occured |= all_paths.list_elements();
      list_paths_occured = true;
      break;
    case append:
      error_occured |= all_paths.append(next_command.arg);
      break;
    case prepend:
      error_occured |= all_paths.prepend(next_command.arg);
      break;
    case remove_first:
      error_occured |= all_paths.remove_first(next_command.arg);
      break;
    case remove_last:
      error_occured |= all_paths.remove_last(next_command.arg);
      break;
    case remove_all:
      error_occured |= all_paths.remove_all(next_command.arg);
      break;
    case before_elm:
        all_paths.set_hook(next_command.arg, before);
      break;
    case after_elm:
      all_paths.set_hook(next_command.arg, after);
      break;
    case use_regex:
      all_paths.set_use_regex(true);
      break;
    case reset_options:
      all_paths.clear_hook();
      all_paths.set_colour(false);
      all_paths.set_use_regex(false);
      break;
    case colour:
        all_paths.set_colour(true);
      break;
    default:
      cout << "Error: Illegal or no command!" << endl;
    }
    commands.pop();
  }

  if (unique_opt) error_occured |= all_paths.uniquify();

  if (!list_paths_occured) {
    cout << all_paths << endl;
  }
  
  return error_occured;
}
